src/MultiAgentReinforcementLearning/src/policies/maddpg.jl

```diff
@@ -0,0 +1,157 @@
+- `traces`, set to `SARTS` if you are apply to an environment of `MINIMAL_ACTION_SET`, or `SLARTSL` if you are to apply to an environment of `FULL_ACTION_SET`.
+- `batch_size::Int`
+- `update_freq::Int`
+- `update_step::Int`, count the step.
+- `rng::AbstractRNG`.
+"""
+mutable struct MADDPGManager{T <: AbstractPolicy} <: AbstractPolicy
+    agents::Dict{<:Any, T}
+    batch_size::Int
+    update_freq::Int
+    update_step::Int
+    rng::AbstractRNG
+end
+
+# used for simultaneous environments.
+function (π::MADDPGManager)(env::AbstractEnv)
+    while current_player(env) == chance_player(env)
+        env |> legal_action_space |> rand |> env
+    end
+    Dict(
+        player => agent(env)
+        for (player, agent) in π.agents)
+end
+
+function (π::MADDPGManager{<: Agent})(::PreActStage, env::AbstractEnv)
+    for (player, agent) in π.agents
+        agent.cache = (agent.cache..., state=state(env, player))
+    end
+    
+end
+
+function (π::MADDPGManager{<: Agent})(::PostActStage, env::AbstractEnv)
+    for (player, agent) in π.agents
+        agent.cache = (agent.cache..., reward=reward(env, player), terminal=is_terminated(env))
+    end
+end
+
+function (π::MADDPGManager)(stage::AbstractStage, env::AbstractEnv)
+    for (_, agent) in π.agents
+        agent(stage, env)
+    end
+end
+
+
+
+# # update policy
+function RLBase.optimise!(π::MADDPGManager)
+    π.update_step % π.update_freq == 0 || return
+    π.update_step += 1
+
+    for (_, agent) in π.agents
+        length(agent.trajectory.container) > agent.policy.update_after || return
+        length(agent.trajectory.container) > π.batch_size || return
+    end
+
+    # get training data
+    batches = Dict((player, [b for b in agent.trajectory])
+                   for (player, agent) in π.agents)
+
+    # do not optimize if one player has empty batch
+    all(length(s[2]) != 0 for s in batches) || return
+    # make sure that all agents have the same amount of batches
+    @assert all(length(s[2]) == length(b[2]) 
+    for (s,b) in Base.Iterators.product(batches, batches) 
+    if s[1] != b[1])
+    
+    # n_batches = length(first(batches)[2]) since all agents
+    # have the same amount of batches. Otherwise, this algorithm
+    # will not be able to be trained accordingly unless dummy actions
+    # are chosen based on the current state of the other agent`s current policy
+    # state
+    for i ∈ 1:length(first(batches)[2])
+      optimise!(π, Dict(player => batches[player][i] for (player, ) in π.agents))
+    end
+    
+end
+
+function RLBase.optimise!(π::MADDPGManager, player_batches)
+    # get s, a, s′ for critic
+    s = vcat((player_batches[player][:state] for (player, _) in π.agents)...)
```

Comment:
Do you have a sense of what this does in terms of creating allocations? Is performance ok? (Naturally, we can merge this and then look at performance in a second step, but it's always good to get the API setup so that performance improvements are easier to implement.

Reply:
Actually, I have not thought about allocation here. To avoid this, it would be necessary that every agent only updates its corresponding parts of the “whole” state. This makes the code a bit scrambled in order of readability. For sure, this is a fact one should have a look regarding performance.